dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java

```diff
@@ -135,23 +162,244 @@ public void testSwitchExpressionsNotSupported() {
     }
   }
 
-  public void testSwitchExpressionsInitializerShouldFail() {
-    try {
-      compileSnippet("void", "    int i = switch(1) {\n" +
-          "      case 1:\n" +
-          "        yield 2;\n" +
-          "      default:\n" +
-          "        yield 7;\n" +
-          "    };");
-      fail("Compile should have failed but succeeded, switch expressions as initializer should fail.");
-    } catch (Exception e) {
-      if (!(e.getCause() instanceof InternalCompilerException)
-          && !(e instanceof InternalCompilerException)) {
-        e.printStackTrace();
-        fail();
-      }
-      assertEquals("Switch expressions not yet supported", e.getMessage());
-    }
+  public void testInstanceOfPatternMatching() throws UnableToCompleteException {
```

Comment:
One more set of tests I can think of that might break with the `X $instanceof_1` decl: a `var` type, for something that can't be expressed easily, like an anonymous inner class, or a type intersection (like was fixed by https://github.com/gwtproject/gwt/commit/ffe61834bf8b83fb3d276b5e4c1f9bdf0af4e9c2) since that was also using `typeMap.get(...)`.

```java
var obj = new SomeClass() {
  public String addAnotherField = "hello";
};
if (obj instanceof SomeClass a) {
  //...
}
```
This could really only matter if there was some other logic going on that would let the type be something else entirely, I'm not thinking about "how this makes sense", but "how this might break the new declaration".

```java
var value = true ? "a" : 'c';
if (obj instanceof CharSequence c) {
  //...
}
```
This one sort of makes more sense, but still is mostly about "hmm, that code looks like something else that has been able to break before".

I'm not sure if this test belongs here or in the Java17Test, up to you.

Reply:
for the first one isnt that a compilation error from the java compiler itself, should we also guard against those, I thought we get those for free?

for the 2nd one I will add the test.

Reply:
```
public class Test {

  public static void main(String[] args) {

    var testPolygon = new Polygon(){
      public int sideWidth = 10;
    };

    if(testPolygon instanceof Polygon p){
      System.out.println(p.sides);
    }
  }
}
```
```
java: expression type <anonymous Polygon> is a subtype of pattern type Polygon
```

Reply:
Ah - if I've made a mistake here and compiling the code catches (with JDT) it, then it wouldn't be a helpful test - I was trying to craft a bug that would only apply to how we're transforming this new instanceof by introducing new locals. We should not need negative test cases for JDT failures (esp as presumably most users are running their code through javac first anyway).

That said, apparently Java 21 can handle the Polygon example, and refers to it as an "unconditional pattern in instanceof" if you try to compile it with release=17, but if you use Java 17 on this, I do see the error you got. Changing the `var` to Object` lets it work in plain Java 17, but is no longer testing what I was concerned about.

The ternary seems to compile in Java 17 though? I can't say that I understand yet why var is okay for one but not the other...